### PR TITLE
Add funcs to add/delete finalizer on ClusterSlice 

### PR DESCRIPTION
**Pull request by panslava**
- Added general functions to patch ClusterSlice ObjectMetadata
- Added functions to add/delete neg cleanup finalizer on ClusterSlice.
  This will allow to properly clean neg controllers when cluster slice
  is deleted

/assign @gauravkghildiyal

### CODE REVIEW

File: pkg/utils/patch/patch.go [hunk 1]
```diff
@@ -84,3 +88,40 @@ func PatchServiceLoadBalancerStatus(client coreclient.CoreV1Interface, svc *core
 	_, err := svchelpers.PatchService(client, svc, newSvc)
 	return err
 }
+
+// PatchClusterSliceObjectMetadata patches the given ClusterSlice's metadata based on new metadata.
+func PatchClusterSliceObjectMetadata(client clustersliceclient.Interface, cs *clusterslice.ClusterSlice, newObjectMetadata metav1.ObjectMeta) error {
+	// Reset Spec to ensure only ObjectMeta is patched.
+	newCS := cs.DeepCopy()
+	newCS.Spec = cs.Spec
```
Comment (gauravkghildiyal) [inline pkg/utils/patch/patch.go:96]: If we deepcopied cs, why are we copying the spec again?

File: pkg/multiproject/finalizer/finalizer.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+	updatedObjectMeta := cs.ObjectMeta.DeepCopy()
+	updatedObjectMeta.Finalizers = slice.RemoveString(updatedObjectMeta.Finalizers, key, nil)
+	logger.V(2).Info("Deleting finalizer from ClusterSlice", "finalizerKey", key, "clusterSlice", fmt.Sprintf("%s/%s", cs.Namespace, cs.Name))
+	return patch.PatchClusterSliceObjectMetadata(csClient, cs, *updatedObjectMeta)
+}
+
+// HasGivenFinalizer is true if the passed in meta has the specified finalizer.
+func HasGivenFinalizer(m metav1.ObjectMeta, key string) bool {
+	return slice.ContainsString(m.Finalizers, key, nil)
```
Comment (gauravkghildiyal) [inline pkg/multiproject/finalizer/finalizer.go:52]: nit: `slices.Contains(...)` can avoid that extra `nil` argument.

File: pkg/multiproject/finalizer/finalizer_test.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+			// Check if the finalizer is present as expected
+			hasFinalizer := slice.ContainsString(updatedCS.Finalizers, ClusterSliceNEGCleanupFinalizer, nil)
+			if hasFinalizer != tc.expectedFinalizerPresent {
+				t.Errorf("Finalizer presence mismatch: expected %v, got %v", tc.expectedFinalizerPresent, hasFinalizer)
+			}
+		})
+	}
+}
+
+func TestEnsureDeleteClusterSliceNEGCleanupFinalizer(t *testing.T) {
```
Comment (gauravkghildiyal) [inline pkg/multiproject/finalizer/finalizer_test.go:78]: nit: Any reason why we have both `EnsureDelete` (like do we want to test both, or a typo?)

**Review by gauravkghildiyal — COMMENTED** [review]
Thanks @panslava!

This LGTM (pending remaining)

File: pkg/utils/patch/patch.go (hunk 1 above)
Comment (panslava) [reply pkg/utils/patch/patch.go:96]: oh yeh, good catch, no reason for that...

File: pkg/multiproject/finalizer/finalizer.go (hunk 2 above)
Comment (panslava) [reply pkg/multiproject/finalizer/finalizer.go]: replaced

File: pkg/multiproject/finalizer/finalizer_test.go (hunk 3 above)
Comment (panslava) [reply pkg/multiproject/finalizer/finalizer_test.go]: Leftover, before I just called the actual function "EnsureDelete" (as done in old code), but then decided there is no reason to add "Ensure"

File: pkg/multiproject/finalizer/finalizer.go (hunk 2 above)
Comment (panslava) [reply pkg/multiproject/finalizer/finalizer.go]: also added one more commit to replace in old code

**Comment by panslava** [comment]
@gauravkghildiyal ready for rereview

File: pkg/multiproject/finalizer/finalizer.go (hunk 2 above)
Comment (gauravkghildiyal) [reply pkg/multiproject/finalizer/finalizer.go]: Thanks for the cleanup!

**Pull request merged**